### PR TITLE
support  TURI_FILEIO_ALTERNATIVE_SSL_CERT_FILE;

### PR DESCRIPTION
**Pull request by guihao-liang**
I have to admit I overlooked this one.

### CODE REVIEW

**Comment by TobyRoseman** [comment]
fixes #3101

**Comment by guihao-liang** [comment]
> It looks like this only works for `TURI_FILEIO_ALTERNATIVE_SSL_CERT_FILE`. We also want it to work for `TURI_FILEIO_ALTERNATIVE_SSL_CERT_DIR`.

Just realized that. It should be good now.

File: src/core/storage/fileio/s3_api.cpp [hunk 1]
```diff
@@ -147,6 +147,20 @@ S3Client init_aws_sdk_with_turi_env(s3url& parsed_url) {
 
   // TODO: add proxy support
   // clientConfiguration.proxyHost = proxy.c_str();
+  if (!get_alternative_ssl_cert_file().empty()) {
+    auto fstaus = get_file_status(get_alternative_ssl_cert_file());
```
Comment (TobyRoseman) [inline src/core/storage/fileio/s3_api.cpp:153]: I don't think this is the right way. I think you should be getting the `clientConfiguration.caPath` value from `get_alternative_ssl_cert_dir()` not from `get_alternative_ssl_cert_file()`.

Also I'm not sure it makes sense but I think the should be possible for both `clientConfiguration.caPath` and `clientConfiguration.caFile` to be set.

**Comment by guihao-liang** [comment]
passed the internal test.

File: src/core/storage/fileio/s3_api.cpp [hunk 2]
```diff
@@ -148,6 +148,27 @@ S3Client init_aws_sdk_with_turi_env(s3url& parsed_url) {
...
   // clientConfiguration.proxyHost = proxy.c_str();
 
+  // set path or file for ssl certs
+  if (!get_alternative_ssl_cert_file().empty()) {
+    auto fstaus = get_file_status(get_alternative_ssl_cert_file());
+    ASSERT_TRUE(fstaus.second.empty());
+    if (fstaus.first == file_status::REGULAR_FILE) {
+      clientConfiguration.caFile = get_alternative_ssl_cert_file().c_str();
+    } else {
+      log_and_throw("Invalid file for alternative SSL certificate");
```
Comment (TobyRoseman) [inline src/core/storage/fileio/s3_api.cpp:158]: Let's give a little more detail in the error message here. Something like:
`Invalid file for alternative SSL certificate. Value of TURI_FILEIO_ALTERNATIVE_SSL_CERT_FILE must be a regular file.`

Including the value of `get_alternative_ssl_cert_file()` in the error message would also be nice.

File: src/core/storage/fileio/s3_api.cpp [hunk 3]
```diff
@@ -148,6 +148,27 @@ S3Client init_aws_sdk_with_turi_env(s3url& parsed_url) {
...
+    }
+  }
+
+  if (!get_alternative_ssl_cert_dir().empty()) {
+    auto fstaus = get_file_status(get_alternative_ssl_cert_dir());
+    ASSERT_TRUE(fstaus.second.empty());
+    if (fstaus.first == file_status::DIRECTORY) {
+      clientConfiguration.caPath = get_alternative_ssl_cert_dir().c_str();
+    } else {
+      log_and_throw("Invalid path for alternative SSL certificate");
```
Comment (TobyRoseman) [inline src/core/storage/fileio/s3_api.cpp:168]: Let give a better error message here too.

File: src/core/storage/fileio/s3_api.cpp (hunk 2 above)
Comment (guihao-liang) [reply src/core/storage/fileio/s3_api.cpp:158]: right, I should give some care to customers.

**Pull request merged**